src/mono/mono/mini/interp/jiterpreter.c

```diff
@@ -1187,6 +1170,44 @@ mono_jiterp_get_member_offset (int member) {
 	}
 }
 
+#define JITERP_COUNTER_TRACE_CANDIDATES 0
+#define JITERP_COUNTER_TRACES_COMPILED 1
+#define JITERP_COUNTER_ENTRY_WRAPPERS_COMPILED 2
+#define JITERP_COUNTER_JIT_CALLS_COMPILED 3
+#define JITERP_COUNTER_DIRECT_JIT_CALLS_COMPILED 4
+#define JITERP_COUNTER_FAILURES 5
+#define JITERP_COUNTER_BYTES_GENERATED 6
+#define JITERP_COUNTER_NULL_CHECKS_ELIMINATED 7
+#define JITERP_COUNTER_NULL_CHECKS_FUSED 8
+#define JITERP_COUNTER_BACK_BRANCHES_EMITTED 9
+#define JITERP_COUNTER_BACK_BRANCHES_NOT_EMITTED 10
+#define JITERP_COUNTER_ELAPSED_GENERATION 11
+#define JITERP_COUNTER_ELAPSED_COMPILATION 12
+#define JITERP_COUNTER_MAX JITERP_COUNTER_ELAPSED_COMPILATION
+
+#define JITERP_COUNTER_UNIT 100
+
+static long counters[JITERP_COUNTER_MAX + 1] = {0};
+
+static long *
+mono_jiterp_get_counter_address (int counter) {
+	g_assert ((counter >= 0) && (counter <= JITERP_COUNTER_MAX));
+	return &counters[counter];
+}
+
+EMSCRIPTEN_KEEPALIVE double
+mono_jiterp_get_counter (int counter) {
+	long actual_result = *(mono_jiterp_get_counter_address (counter));
+	return ((double)actual_result) / JITERP_COUNTER_UNIT;
+}
+
+EMSCRIPTEN_KEEPALIVE double
+mono_jiterp_modify_counter (int counter, double delta) {
+	long actual_delta = (long)(delta * JITERP_COUNTER_UNIT);
+	long actual_result = atomic_fetch_add ((atomic_long *)mono_jiterp_get_counter_address (counter), actual_delta);
```

Comment:
is it worth having `#ifdef DISABLE_THREADS` ?

Reply:
I thought not, but you're right that it could make the binary larger by pulling in stdatomic implementation in single-thread builds, so I'll ifdef it.